File: frontend-client/src/lib/hanaro.ts
// 하나로 SSO 공직자 판별 API 호출 + Clerk 유저 매핑 헬퍼 (서버 전용)
import { clerkClient } from "@clerk/nextjs/server";

export type OfficialCheckResult =
  | { ok: true; isOfficial: boolean }
  | { ok: false; reason: "invalid_credentials" | "rate_limited" | "server_config" | "upstream_error" };

const DEFAULT_URL = "https://hanaro.ffwp.or.kr/API_kim/officialLoginCheck2";

// 하나로 SSO 아이디/비밀번호를 서버에서 검증하고 공직자 여부를 받는다.
// keyValue/비밀번호는 절대 로깅하지 않는다.
export async function checkOfficial(userid: string, password: string): Promise<OfficialCheckResult> {
  const key = process.env.OFFICIAL_CHECK_KEY;
  const url = process.env.OFFICIAL_CHECK_URL ?? DEFAULT_URL;
  if (!key) return { ok: false, reason: "server_config" };

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 8000);
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({ keyValue: key, userid, password }),
      signal: controller.signal,
      cache: "no-store",
    });
    if (res.status === 429) return { ok: false, reason: "rate_limited" };
    const data = (await res.json().catch(() => null)) as
      | { authenticated?: boolean; isOfficial?: boolean; error?: string }
      | null;
    if (!data) return { ok: false, reason: "upstream_error" };
    if (data.error === "rate_limited") return { ok: false, reason: "rate_limited" };
    if (data.error === "invalid_key" || data.error === "missing_parameter")
      return { ok: false, reason: "server_config" };
    if (data.authenticated === true) return { ok: true, isOfficial: data.isOfficial === true };
    return { ok: false, reason: "invalid_credentials" };
  } catch {
    return { ok: false, reason: "upstream_error" };
  } finally {
    clearTimeout(timer);
  }
}

// 하나로 userid를 Clerk 유저에 idempotent하게 매핑한다(externalId 기준). isOfficial은 매 로그인 갱신.
export async function findOrCreateHanaroUser(userid: string, isOfficial: boolean): Promise<string> {
  const client = await clerkClient();
  const extId = `hanaro:${userid}`;
  const existing = await client.users.getUserList({ externalId: [extId], limit: 1 });
  if (existing.totalCount > 0) {
    const u = existing.data[0];
    await client.users.updateUserMetadata(u.id, {
      publicMetadata: { ...u.publicMetadata, isOfficial, provider: "hanaro", hanaroUserId: userid },
    });
    return u.id;
  }
  try {
    const created = await client.users.createUser({
      externalId: extId,
      emailAddress: [`${userid}@hanaro.local`],
      skipPasswordRequirement: true,
      publicMetadata: { isOfficial, provider: "hanaro", hanaroUserId: userid },
    });
    return created.id;
  } catch (e) {
    // 동시 최초 로그인 race — 이미 생성됐으면 재조회
    const retry = await client.users.getUserList({ externalId: [extId], limit: 1 });
    if (retry.totalCount > 0) return retry.data[0].id;
    throw e;
  }
}

File: backend/app/core/hanaro.py
# 하나로 SSO 공직자 판별 API v2 를 호출해 로그인 검증과 공직자 여부를 받는 클라이언트

import logging
from dataclasses import dataclass

import httpx

from app.core.config import get_settings

logger = logging.getLogger(__name__)

# 규격서(공직자 판별 API v2, 2026-07-16) 2장 — 엔드포인트
DEFAULT_URL = "https://hanaro.ffwp.or.kr/API_kim/officialLoginCheck2"
TIMEOUT_SEC = 8.0


@dataclass(frozen=True)
class OfficialCheckResult:
    """하나로 검증 결과.

    reason 은 실패했을 때만 채워진다.
    - invalid_credentials  아이디/비밀번호 불일치 (사용자 잘못)
    - rate_limited         아이디당 실패 15회 누적 (규격서 5장)
    - server_config        우리 서버에 발급 키가 없음 (우리 잘못)
    - upstream_key_rejected 키를 보냈으나 하나로가 거부 (IT팀 확인 필요)
    - upstream_error       네트워크·응답 형식 오류
    """

    ok: bool
    is_official: bool = False
    reason: str | None = None


async def check_official(userid: str, password: str) -> OfficialCheckResult:
    """하나로 아이디·비밀번호를 검증하고 공직자 여부를 받는다.

    발급 키와 비밀번호는 어떤 경우에도 로깅하지 않는다(규격서 8장).
    """
    settings = get_settings()
    key = (
        settings.OFFICIAL_CHECK_KEY.get_secret_value()
        if settings.OFFICIAL_CHECK_KEY
        else None
    )
    url = settings.OFFICIAL_CHECK_URL or DEFAULT_URL

    if not key:
        logger.error("OFFICIAL_CHECK_KEY 미설정 — 하나로 API 를 호출하지 않는다")
        return OfficialCheckResult(ok=False, reason="server_config")

    try:
        async with httpx.AsyncClient(timeout=TIMEOUT_SEC) as client:
            resp = await client.post(
                url,
                data={"keyValue": key, "userid": userid, "password": password},
            )
    except Exception as e:
        logger.error("하나로 API 호출 실패: %s — %s", type(e).__name__, e)
        return OfficialCheckResult(ok=False, reason="upstream_error")

    # 규격서 5장 — 아이디당 실패 15회 누적 시 429. 마지막 실패로부터 5분 뒤 자동 해제.
    if resp.status_code == 429:
        return OfficialCheckResult(ok=False, reason="rate_limited")

    try:
        data = resp.json()
    except ValueError:
        logger.error("하나로 응답 JSON 파싱 실패 (status=%s)", resp.status_code)
        return OfficialCheckResult(ok=False, reason="upstream_error")

    error = data.get("error")
    if error == "rate_limited":
        return OfficialCheckResult(ok=False, reason="rate_limited")
    if error in ("invalid_key", "missing_parameter"):
        # 업스트림 에러코드만 남긴다 — 설정 오류를 진단할 유일한 단서다.
        # (v1 전용 키를 v2 주소에 쓰면 invalid_key 가 온다.)
        logger.error("하나로가 설정을 거절: %s (url=%s)", error, url)
        return OfficialCheckResult(ok=False, reason="upstream_key_rejected")

    # 규격서 4장 — isOfficial 은 authenticated 가 true 일 때만 유효하다.
    if data.get("authenticated") is True:
        return OfficialCheckResult(ok=True, is_official=data.get("isOfficial") is True)

    return OfficialCheckResult(ok=False, reason="invalid_credentials")
